### PR TITLE
GHA: improve vcpkg cache, add BoringSSL ECH and LibreSSL MSVC jobs

### PR DESCRIPTION
**Pull request by vszakats**
- cache on a per-package basis.
  Replace manual caching with a built-in solution. It shares cached
  package builds between jobs, e.g. libssh2 only builds once
  per platform (instead of once per job). Individual packages are built
  as needed (not the whole per-job tree). It also fixes the duplicate
  cache entry issues.
  Ref: https://learn.microsoft.com/en-us/vcpkg/consume/binary-caching-github-actions-cache
  Follow-up to e26cbe20cbedbea0ca743dd33880517309315cb2 #13979
  Follow-up to cb22cfca69bded45bf7f9c72c8e6764990490f11 #14077

- add BoringSSL job with ECH enabled. The first such job in the curl CI.

- add LibreSSL job.

- use vcpkg pre-installed on the runner image, instead of rolling our
  own. This is quicker, simpler and more robust.
  Follow-up to e26cbe20cbedbea0ca743dd33880517309315cb2 #13979

- show pre-installed vcpkg and ports version.

- drop `gsasl` dependency till it reaches the pre-installed vcpkg ports.

- re-add `find .` to see the binaries generated.

- simplify setting up `PATH`.

- exclude failing tests for any job enabling WinIDN.

- drop collecting and uploading log archives. We already dump CMake
  logs, and our build doesn't use Ninja. Rest of files weren't generated
  by the curl build. We don't aim to debug vcpkg package builds.

Closes #14090

---

TODO:
- [x] rebase on #14077.
- [x] squash into meaningful sub-commits.
- [ ] delete cache entries after merge.

### CODE REVIEW

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -40,442 +40,62 @@ concurrency:
...
+    env:
+      VCPKG_BINARY_SOURCES: 'clear;x-gha,readwrite'
+      VCPKG_DISABLE_METRICS: '1'
     strategy:
       matrix:
         include:
           # name should be unique and without special chars. It creates a folder with that name for the cache.
         - name: 'schannel'
-          install: 'zstd zlib nghttp2 libssh2[core,zlib]'
+          install: 'zstd zlib brotli libpsl[libidn2] nghttp2 libssh2[core,zlib]'
```
Comment (vszakats) [inline .github/workflows/windows.yml:55]:
```suggestion
          install: 'zstd zlib brotli libpsl[core,libidn2] nghttp2 libssh2[core,zlib]'
```

**Comment by vszakats** [comment]
Hit this with LibreSSL from vcpkg. There is some confusion with also picking up a non-vcpkg OpenSSL, may be related:
```
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol sk_num referenced in function Curl_ossl_verifyhost [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol sk_value referenced in function Curl_ossl_verifyhost [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol sk_pop_free referenced in function cert_stuff [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol sk_pop referenced in function cert_stuff [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol EVP_MD_CTX_create referenced in function Curl_sha512_256it [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol EVP_MD_CTX_destroy referenced in function Curl_sha512_256it [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol SSL_CTX_add0_chain_cert referenced in function SSL_CTX_use_certificate_chain_blob [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol SSL_CTX_clear_chain_certs referenced in function SSL_CTX_use_certificate_chain_blob [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol SSL_CTX_set1_groups_list referenced in function Curl_ossl_ctx_init [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol SSL_CTX_set_min_proto_version referenced in function ossl_set_ssl_version_min_max [D:\a\curl\curl\bld\src\curl.vcxproj]
libcurl-d.lib(unity_0_c.obj) : error LNK2019: unresolved external symbol SSL_CTX_set_max_proto_version referenced in function ossl_set_ssl_version_min_max [D:\a\curl\curl\bld\src\curl.vcxproj]
D:\a\curl\curl\bld\src\curl.exe : fatal error LNK1120: 11 unresolved externals [D:\a\curl\curl\bld\src\curl.vcxproj]
```
Ref: https://github.com/curl/curl/actions/runs/9787764034/job/27024766559#step:6:41

**Comment by talregev** [comment]
You can have hints in the vcpkg patches.

**Comment by vszakats** [comment]
> You can have hints in the vcpkg patches.

Do you have a specific place to look? (I checked the LibreSSL package, which says it's incompatible with OpenSSL, but the build phase is fine, the issue comes when trying to build against curl. Perhaps a vcpkg curl patch/wrapper?)

**Comment by talregev** [comment]
vcpkg curl patches.
one hint:
Curl should add the decencies of the packages.

**Pull request closed without merging**